connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala

```diff
@@ -0,0 +1,212 @@
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+class ProtobufCatalystDataConversionSuite
+    extends SparkFunSuite
+    with SharedSparkSession
+    with ExpressionEvalHelper {
+
+  private def checkResult(
+      data: Literal,
+      descFilePath: String,
+      messageName: String,
+      expected: Any): Unit = {
+    checkEvaluation(
+      ProtobufDataToCatalyst(
+        CatalystDataToProtobuf(data, descFilePath, messageName),
+        descFilePath,
+        messageName,
+        Map.empty),
+      prepareExpectedResult(expected))
+  }
+
+  protected def checkUnsupportedRead(
+      data: Literal,
+      descFilePath: String,
+      actualSchema: String,
+      badSchema: String): Unit = {
+
+    val binary = CatalystDataToProtobuf(data, descFilePath, actualSchema)
+
+    intercept[Exception] {
+      ProtobufDataToCatalyst(binary, descFilePath, badSchema, Map("mode" -> "FAILFAST")).eval()
+    }
+
+    val expected = {
+      val expectedSchema = ProtobufUtils.buildDescriptor(descFilePath, badSchema)
+      SchemaConverters.toSqlType(expectedSchema).dataType match {
+        case st: StructType =>
+          Row.fromSeq((0 until st.length).map { _ =>
+            null
+          })
+        case _ => null
+      }
+    }
+
+    checkEvaluation(
+      ProtobufDataToCatalyst(binary, descFilePath, badSchema, Map("mode" -> "PERMISSIVE")),
+      expected)
+  }
+
+  protected def prepareExpectedResult(expected: Any): Any = expected match {
+    // Spark byte and short both map to Protobuf int
+    case b: Byte => b.toInt
+    case s: Short => s.toInt
+    case row: GenericInternalRow => InternalRow.fromSeq(row.values.map(prepareExpectedResult))
+    case array: GenericArrayData => new GenericArrayData(array.array.map(prepareExpectedResult))
+    case map: MapData =>
+      val keys = new GenericArrayData(
+        map.keyArray().asInstanceOf[GenericArrayData].array.map(prepareExpectedResult))
+      val values = new GenericArrayData(
+        map.valueArray().asInstanceOf[GenericArrayData].array.map(prepareExpectedResult))
+      new ArrayBasedMapData(keys, values)
+    case other => other
+  }
+
+  private val testingTypes = Seq(
+    StructType(StructField("int32_type", IntegerType, nullable = true) :: Nil),
+    StructType(StructField("double_type", DoubleType, nullable = true) :: Nil),
+    StructType(StructField("float_type", FloatType, nullable = true) :: Nil),
+    StructType(StructField("bytes_type", BinaryType, nullable = true) :: Nil),
+    StructType(StructField("string_type", StringType, nullable = true) :: Nil))
+
+  private val catalystTypesToProtoMessages: Map[DataType, String] = Map(
+    IntegerType -> "IntegerMsg",
+    DoubleType -> "DoubleMsg",
+    FloatType -> "FloatMsg",
+    BinaryType -> "BytesMsg",
+    StringType -> "StringMsg")
+
+  testingTypes.foreach { dt =>
+    val seed = 1 + scala.util.Random.nextInt((1024 - 1) + 1)
```

Comment:
this one fails when seed = 38, see https://github.com/MaxGekk/spark/actions/runs/3391675654/jobs/5638519849

Reply:
@MaxGekk This problem was solved by #38286. Was your fork in sync with master? I can open a JIRA if it fails again. I repeated the test with the above seed = 38. There were no failures.

```
[info] ProtobufCatalystDataConversionSuite:
[info] - single StructType(StructField(int32_type,IntegerType,true)) with seed 38 (67 milliseconds)
[info] - single StructType(StructField(double_type,DoubleType,true)) with seed 38 (22 milliseconds)
[info] - single StructType(StructField(float_type,FloatType,true)) with seed 38 (18 milliseconds)
[info] - single StructType(StructField(bytes_type,BinaryType,true)) with seed 38 (23 milliseconds)
[info] - single StructType(StructField(string_type,StringType,true)) with seed 38 (27 milliseconds)
[info] - Handle unsupported input of message type (28 milliseconds)
[info] - filter push-down to Protobuf deserializer (13 milliseconds)
[info] - ProtobufDeserializer with binary type (0 milliseconds)
[info] - Full names for message using descriptor file (1 millisecond)
```

Reply:
This happened on the recent master. The commit is [216a5d2c39ea3d9e869818692b2f4c0f2652aa56](https://github.com/apache/spark/commit/216a5d2c39ea3d9e869818692b2f4c0f2652aa56):

```
$ git diff
```
```diff
diff --git a/connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala b/connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala
index 271c5b0fec..080bf1eb1f 100644
--- a/connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala
+++ b/connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala
@@ -123,7 +123,7 @@ class ProtobufCatalystDataConversionSuite
     StringType -> ("StringMsg", ""))

   testingTypes.foreach { dt =>
-    val seed = 1 + scala.util.Random.nextInt((1024 - 1) + 1)
+    val seed = 38
     test(s"single $dt with seed $seed") {

       val (messageName, defaultValue) = catalystTypesToProtoMessages(dt.fields(0).dataType)
```

```
$ build/sbt "test:testOnly *ProtobufCatalystDataConversionSuite"
```
```
[info] - single StructType(StructField(double_type,DoubleType,true)) with seed 38 *** FAILED *** (10 milliseconds)
[info]   java.lang.NullPointerException:
[info]   at org.apache.spark.sql.protobuf.ProtobufCatalystDataConversionSuite.$anonfun$new$2(ProtobufCatalystDataConversionSuite.scala:134)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
```

Reply:
Here is the JIRA for the bug: https://issues.apache.org/jira/browse/SPARK-41015

Reply:
@MaxGekk thank you will look into it.